tripy/tripy/flat_ir/flat_ir.py

```diff
@@ -31,83 +34,160 @@ class FlatIR:
-                                for mlir_out, flatir_out in zip(layer_outputs, op.outputs):
-                                    type = get_op_result_or_value(mlir_out).type
-                                    flatir_out.shape = tuple(
-                                        [
-                                            (-1 if type.is_dynamic_dim(i) else type.get_dim_size(i))
-                                            for i in range(type.rank)
-                                        ]
-                                    )
-
-                            ops.extend(layer_outputs)
-                            mlir_ops.update(zip([out.name for out in op.outputs], layer_outputs))
-
-                        func_dialect.ReturnOp([mlir_ops[o.name] for o in self.outputs])
+                mlir_module = ir.Module.create()
+                with ir.InsertionPoint(mlir_module.body) as insertion_point:
+                    mlir_tensor_map: Dict[str, Union[ir.Operation, ir.Value]] = {}
+
+                    # Create private functions for each FlatIRFunction
+                    for flat_ir_func in self.functions.values():
+                        func_input_types = [input_tensor.to_mlir() for input_tensor in flat_ir_func.inputs]
+                        func_output_types = [output_tensor.to_mlir() for output_tensor in flat_ir_func.outputs]
+                        func_type = ir.FunctionType.get(func_input_types, func_output_types)
+                        func_op = func_dialect.FuncOp(
+                            flat_ir_func.name, func_type, ip=insertion_point, visibility="private"
+                        )
+
+                        with ir.InsertionPoint(func_op.add_entry_block()):
+                            # Map function inputs to MLIR tensor map
+                            for index, input_tensor in enumerate(flat_ir_func.inputs):
+                                mlir_tensor_map[input_tensor.name] = func_op.arguments[index]
+
+                            # Process each operation in the function
+                            for operation in flat_ir_func.ops:
+                                operation_inputs = [
+                                    mlir_tensor_map[input_tensor.name] for input_tensor in operation.inputs
+                                ]
+                                operation_outputs = operation.to_mlir(operation_inputs)
+                                for mlir_output, flat_ir_output in zip(operation_outputs, operation.outputs):
+                                    mlir_tensor_map[flat_ir_output.name] = mlir_output
+
+                            # Add return operation for the function
+                            func_dialect.ReturnOp(
+                                [mlir_tensor_map[output_tensor.name] for output_tensor in flat_ir_func.outputs]
+                            )
+
+                    # Create the main function
```

Comment:
Is it possible to add the main function to `self.functions` so that we don't need a special case for it?

Reply:
It should be possible. We apply some post-processing for the main function type for dynamic shapes. I am first getting this to work end to end with dynamic shapes. I will try your suggestion afterward.